lib/srv/db/common/auth.go

```diff
@@ -838,29 +838,43 @@ func (a *dbAuth) GetAWSIAMCreds(ctx context.Context, sessionCtx *Session) (strin
 	dbAWS := sessionCtx.Database.GetAWS()
 	awsAccountID := dbAWS.AccountID
 
-	// If AWS account ID is empty, we need to fetch it from the agent identity.
-	// It has to be filled in case the database username is a partial ARN.
 	if awsutils.IsPartialRoleARN(sessionCtx.DatabaseUser) && awsAccountID == "" {
-		a.cfg.Log.Debugf("Fetching AWS Account ID to build role ARN")
-		stsClient, err := a.cfg.Clients.GetAWSSTSClient(ctx, dbAWS.Region)
-		if err != nil {
-			return "", "", "", trace.Wrap(err)
-		}
+		switch {
+		case dbAWS.AssumeRoleARN != "":
+			a.cfg.Log.Debugf("Using AWS Account ID from assumed role")
+			assumeRoleARN, err := awsutils.ParseRoleARN(dbAWS.AssumeRoleARN)
+			if err != nil {
+				return "", "", "", trace.Wrap(err)
+			}
 
-		identity, err := awslib.GetIdentityWithClient(ctx, stsClient)
-		if err != nil {
-			return "", "", "", trace.Wrap(err)
-		}
+			awsAccountID = assumeRoleARN.AccountID
+		default:
+			a.cfg.Log.Debugf("Fetching AWS Account ID to build role ARN")
+			stsClient, err := a.cfg.Clients.GetAWSSTSClient(ctx, dbAWS.Region)
+			if err != nil {
+				return "", "", "", trace.Wrap(err)
+			}
 
-		awsAccountID = identity.GetAccountID()
+			identity, err := awslib.GetIdentityWithClient(ctx, stsClient)
+			if err != nil {
+				return "", "", "", trace.Wrap(err)
+			}
+
+			awsAccountID = identity.GetAccountID()
+		}
 	}
 
 	arn, err := awsutils.BuildRoleARN(sessionCtx.DatabaseUser, dbAWS.Region, awsAccountID)
 	if err != nil {
 		return "", "", "", trace.Wrap(err)
 	}
 
-	sess, err := a.cfg.Clients.GetAWSSession(ctx, dbAWS.Region, cloud.WithAssumeRole(arn, sessionCtx.Database.GetAWS().ExternalID))
+	baseSession, err := a.cfg.Clients.GetAWSSession(ctx, dbAWS.Region, cloud.WithAssumeRoleFromAWSMeta(dbAWS))
+	if err != nil {
+		return "", "", "", trace.Wrap(err)
+	}
+
+	sess, err := a.cfg.Clients.GetAWSSession(ctx, dbAWS.Region, cloud.WithChainedAssumeRole(baseSession, arn, dbAWS.ExternalID))
```

Comment:
Thanks for updating this!

One nit comment. So when both `assume_role_arn` and `external_id` are specified, `external_id` is only used for the `assume_role_arn` and the second/chained assume role should use empty for external id. 

We had a debate on this as this behaviour is kinda confusing. But having two external ids can be confusing too. The current logic basically assumes `external_id` should be used for the first assume-role, and the second assume-role shouldn't need the external id as they should be in the same account.

https://github.com/gravitational/teleport/blob/f4385c16df26c4e140371cf15f770ff7bfd75158/lib/srv/db/dynamodb/engine.go#L214-L216

Reply:
Done. I've added some tests to ensure the behavior too.